[Concept]: Basics (#366)
exercism/powershell: concepts/basics/about.md

# About

> PowerShell is a dynamically implicit typed (also support strongly typed variables) scripting language from Microsoft, mostly focus on task automation and confifuration management.

Reviewer: ```suggestion
PowerShell is a dynamic and implicitly typed (_but also supporting [strongly typed][strong typed] variables_) scripting language from Microsoft, mostly focused on task automation and configuration management.
```

> PowerShell supports imperative, functional, object-oriented programming, and even pipeline style.

Reviewer: ```suggestion
PowerShell supports imperative, functional, object-oriented, and even pipeline style programming.
```

PowerShell was built on the .NET platform and first released in 2006.

> It has been included by default in all modern Windows OS, and can be installed on Mac and Linux.

Reviewer: ```suggestion
It has been included by default in all modern Windows OS, and can also be installed on Mac and Linux.
```

## Variables

There are two common ways to define a variable in PowerShell.

> Varible name in PowerShell must start with `$` and using the assignment operator `=` to bind the variable name to a value.

Reviewer: ```suggestion
Variable names in PowerShell must start with a `$`, and use the assignment operator `=` to bind the variable name to a value.
```

```powershell
> $intVar = 1
> $doubleVar = 2.0
> $stringVar = "Hello"

# Checking a variable type using the GetType() method
> Write-Host $doubleVar.GetType()
System.Double
```

> You can also explicitly specify a variable's type if you want to enforce a specific type.

Reviewer: ```suggestion
You can also specify a variable's type explicitly, if you want to enforce/require the value be of a certain type.
```

PowerShell will try its best to convert the value from one type to another if possible.

```powershell
# String represents a floating number can be converted to integer
> [int]$staticOne = "1.234"
> $staticOne
1

# String of letters can not be converted to double type
> [double]$staticTwo = "hello"
MetadataError: Cannot convert value "hello" to type "System.Double"
```

### Constants

> Constants in PowerShell are read-only variables that can't be **removed** after defined.

Reviewer 2: This implies they can have their value changed, just not removed.

Author: I thought the read-only mention already addressed that? But yeah I could change the wording here

> Constants in PowerShell are read-only variables that can't be **removed** after defined.

Reviewer: ```suggestion
Constants in PowerShell are read-only variables that can't be **removed** after definition.
```

> This lead us to the second way to define variables using the cmdlet `New-Variable`.

Reviewer: ```suggestion
This leads us to the second way of defining variables: using the cmdlet `New-Variable`.
```

```powershell
> New-Variable -Name "MyConstant" -Value 100 -Option Constant
> $MyConstant
100
```

> For a more detailed and advanced way to create a new variable with various options, please read [New-Variable][New-Variable].

Reviewer: ```suggestion
For more detailed and advanced ways to create a new variable, please read [New-Variable][New-Variable].
```

## Functions

Functions in PowerShell are defined using the `function` keyword.

> Each function can have zero or more formal parameters in `()` parenthesis, followed by its own scope start  with `{` and end with `}`. 

Reviewer 2: ```suggestion
Each function can have zero or more formal parameters in `()` parenthesis, followed by its own scope start with `{` and end with `}`. 
```

> Each function can have zero or more formal parameters in `()` parenthesis, followed by its own scope start  with `{` and end with `}`. 

Reviewer: ```suggestion
Each function can have zero or more formal parameters in `()` parenthesis, followed by the function scope starting  with `{` and ending with `}`. 
```

```powershell
function Greeting {
    return "Hello!"
}
```

> Functions in PowerShell can return values explicitly via the `return` keyword, after being called using the syntax `FuncName $Arg`.

Reviewer: Is there an implicit return when `return` is omitted?

```powershell
function AddTwoNumbers($NumOne, $NumTwo):
    $result = $NumOne + $NumTwo
    return $result

> AddTwoNumbers 10 20
30
```

## Comments

> Single comment in PowerShell start with a `#`, and end at line termination.

Reviewer: ```suggestion
Single comments in PowerShell start with a `#`, and end at line termination.
```

> Comment block in PowerShell start with `<#` and end with `#>`.

Reviewer: ```suggestion
Comment blocks in PowerShell start with `<#` and end with `#>`.
```

```powershell
function Main() {
    #This is a comment
}

> ```

Reviewer: Feels like there should also be a mulit-line example here, to demonstrate what that looks like.

> ```

Reviewer: Feels like there should also be a mulit-line example here, to demonstrate what that looks like.
